[mlir][gpu] Add 'cluster_size' attribute to gpu.subgroup_reduce

andfau-amd opened the pull request:
This enables performing several reductions in parallel, each smaller than the size of the subgroup.

One potential application is flash attention with subgroup-wide matrix multiplication and reduction combined in one kernel. The multiplication operation requires a 2D matrix to be distributed over the lanes of the subgroup, which then constrains the shape the following reduction can have if we want to keep data in registers.

andfau-amd commented:
For broader motivation please see https://github.com/iree-org/iree/issues/18142.

kuhar commented on mlir/include/mlir/Dialect/GPU/IR/GPUOps.td:
```diff
@@ -1208,11 +1208,21 @@ def GPU_SubgroupReduceOp : GPU_Op<"subgroup_reduce", [SameOperandsAndResultType]
+    If a `cluster_size` is not provided, the reduction covers all lanes of the
+    subgroup and the result is equal for all lanes.
+
+    If a `cluster_size` is provided, the subgroup is divided into clusters of
+    `cluster_size` contiguous lanes each, a reduction is done for all lanes of
```
It would be nice to clarify that these clusters are contagious and adjacent (instead of, say, strided / interleaved).

andfau-amd replied:
I did already say the lanes are contiguous. How would I need to word it to add further clarity?

kuhar replied:
Ah, right

kuhar commented on mlir/include/mlir/Dialect/GPU/IR/GPUOps.td:
```diff
@@ -1222,12 +1232,32 @@ def GPU_SubgroupReduceOp : GPU_Op<"subgroup_reduce", [SameOperandsAndResultType]
+    OpBuilder<(ins "Value":$value,
+               "::mlir::gpu::AllReduceOperation":$op,
+               "bool":$uniform), [{
+      build($_builder, $_state, value, op, uniform, /* cluster_size: */ nullptr);
```
to follow the llvm coding style:
```suggestion
      build($_builder, $_state, value, op, uniform, /*cluster_size=*/ nullptr);
```

kuhar commented on mlir/include/mlir/Dialect/GPU/IR/GPUOps.td:
```diff
@@ -1222,12 +1232,32 @@ def GPU_SubgroupReduceOp : GPU_Op<"subgroup_reduce", [SameOperandsAndResultType]
+               "::mlir::gpu::AllReduceOperation":$op,
+               "bool":$uniform,
+               "std::optional<uint32_t>":$cluster_size), [{
+      if (cluster_size)
+        build($_builder, $_state, value, op, uniform, $_builder.getI32IntegerAttr(*cluster_size));
+      else
+        build($_builder, $_state, value, op, uniform, nullptr);
```
I think it might be a bit safer to turn this into a ternary operator so that we don't have to assume how the autogenerated code around this will look like and keep it as a single expression.

andfau-amd replied:
I could do this:

```cpp
      IntegerAttr clusterSize;
      if (cluster_size)
        clusterSize = $_builder.getI32IntegerAttr(*cluster_size);
      build($_builder, $_state, value, op, uniform, clusterSize);
```

I'm not sure its an improvement though.

kuhar replied:
I was thinking of something like:
```
build($_builder, $_state, value, op, uniform, cluster_size ? $_builder.getI32IntegerAttr(*cluster_size) : IntegerAttr{});
```

kuhar commented on mlir/lib/Dialect/GPU/IR/GPUDialect.cpp:
```diff
@@ -620,6 +620,14 @@ LogicalResult gpu::SubgroupReduceOp::verify() {
+
+  if (auto clusterSize = getClusterSize()) {
+    uint32_t size = *clusterSize;
+    if (size < 1 || !llvm::isPowerOf2_32(size)) {
```
`isPowerOf2` already covers being positive

Groverkss replied:
unsigned also covers positive, although im not sure if using unsigned to indicate positiveness is preferred.

kuhar replied:
unsigned is non-negative, not positive

kuhar commented on mlir/lib/Dialect/GPU/Transforms/SubgroupReduceLowering.cpp:
```diff
@@ -183,6 +190,13 @@ struct ScalarSubgroupReduceToShuffles final
+      return op.emitError()
+             << "cluster size " << *clusterSize
+             << " is greater than subgroup size " << subgroupSize;
+    unsigned effectiveClusterSize = clusterSize ? *clusterSize : subgroupSize;
```
We can use `clusterSize.value_or(subgroup_size)`

kuhar commented on mlir/lib/Dialect/GPU/Transforms/SubgroupReduceLowering.cpp:
```diff
@@ -237,6 +253,13 @@ struct VectorSubgroupReduceToShuffles final
+      return op.emitError()
+             << "cluster size " << *clusterSize
+             << " is greater than subgroup size " << subgroupSize;
+    unsigned effectiveClusterSize = clusterSize ? *clusterSize : subgroupSize;
```
Also here

kuhar reviewed (commented):
Thanks for the PR. Would you be able to add some example use case that we could add to the PR description? IIRC this is intended for flash attention?

kuhar commented on mlir/test/Dialect/GPU/subgroup-reduce-lowering.mlir:
```diff
@@ -144,6 +186,33 @@ gpu.module @kernels {
+    // CHECK-SHFL-DAG: %[[C2:.+]] = arith.constant 2 : i32
+    // CHECK-SHFL-DAG: %[[C32:.+]] = arith.constant 32 : i32
+
+    // CHECK-SHFL: %[[V0:.+]] = vector.bitcast %[[ARG0]] : vector<2xf16> to vector<1xi32>
+    // CHECK-SHFL: %[[I0:.+]] = vector.extract %[[V0]][0] : i32 from vector<1xi32>
+    // CHECK-SHFL: %[[S0:.+]], %{{.+}} = gpu.shuffle xor %[[I0]], %[[C1]], %[[C32]] : i32
+    // CHECK-SHFL: %[[BR0:.+]] = vector.broadcast %[[S0]] : i32 to vector<1xi32>
+    // CHECK-SHFL: %[[BC0:.+]] = vector.bitcast %[[BR0]] : vector<1xi32> to vector<2xf16>
+    // CHECK-SHFL: %[[ADD0:.+]] = arith.addf %[[ARG0]], %[[BC0]] : vector<2xf16>
+    // CHECK-SHFL: %[[BC1:.+]] = vector.bitcast %[[ADD0]] : vector<2xf16> to vector<1xi32>
+    // CHECK-SHFL: %[[I1:.+]] = vector.extract %[[BC1]][0] : i32 from vector<1xi32>
+    // CHECK-SHFL: %[[SL:.+]], %{{.+}} = gpu.shuffle xor %{{.+}}, %[[C2]], %[[C32]] : i32
+    // CHECK-SHFL: %[[BRL:.+]] = vector.broadcast %[[SL]] : i32 to vector<1xi32>
+    // CHECK-SHFL: %[[BCL:.+]] = vector.bitcast %[[BRL]] : vector<1xi32> to vector<2xf16>
+    // CHECK-SHFL: %[[ADDL:.+]] = arith.addf %{{.+}}, %[[BCL]] : vector<2xf16>
+    // CHECK-SHFL: "test.consume"(%[[ADDL]]) : (vector<2xf16>) -> ()
```
Do we care about the exact lowering here, or  would it be enough to check that we emit N `shuffles xor`s?

kuhar commented on mlir/include/mlir/Dialect/GPU/IR/GPUOps.td:
````diff
@@ -1208,11 +1208,21 @@ def GPU_SubgroupReduceOp : GPU_Op<"subgroup_reduce", [SameOperandsAndResultType]
     ```mlir
     %1 = gpu.subgroup_reduce add %a : (f32) -> (f32)
     %2 = gpu.subgroup_reduce add %b : (vector<4xf16>) -> (vector<4xf16>)
+    %3 = gpu.subgroup_reduce add %c cluster_size(4) : (f32) -> (f32)
````
nit: we don't need the parens in the result type in any of these examples

kuhar commented on mlir/lib/Dialect/GPU/IR/GPUDialect.cpp:
```diff
@@ -629,6 +637,10 @@ OpFoldResult gpu::SubgroupReduceOp::fold(FoldAdaptor /*adaptor*/) {
     return getResult();
   }
 
+  if (getClusterSize() == 1) {
+    return getValue();
+  }
```
Wouldn't it be better to check for this before uniformity? This seems like a cheaper way of doing things because we won't ever have to fold twice.

Groverkss replied:
nit: drop braces

andfau-amd commented:
> Thanks for the PR. Would you be able to add some example use case that we could add to the PR description? IIRC this is intended for flash attention?

I added something now to the PR description and commit message. I hope it's not too vague.

kuhar commented on mlir/lib/Dialect/GPU/IR/GPUDialect.cpp:
```diff
@@ -620,10 +620,21 @@ LogicalResult gpu::SubgroupReduceOp::verify() {
+  if (auto clusterSize = getClusterSize()) {
+    uint32_t size = *clusterSize;
+    if (!llvm::isPowerOf2_32(size)) {
+      return emitError() << "cluster size " << size << " is not a power of two";
```
```suggestion
      return emitOpError() << "cluster size " << size << " is not a power of two";
```

To print the op inline with the error message

andfau-amd replied:
I've made this change, but to be honest I'm not sure why this is better. It shows the line where the error happened either way, right?

kuhar replied:
I find op errors to be easier to work with because in real models the line number loc doesn't really match anything useful (high level ops get decomposed into dozens/hundreds of simpler instructions all with the same location)

andfau-amd replied:
I see, thanks.

kuhar reviewed (approved):
LGTM

andfau-amd commented on mlir/include/mlir/Dialect/GPU/IR/GPUOps.td:
```diff
@@ -1198,21 +1198,31 @@ def GPU_SubgroupReduceOp : GPU_Op<"subgroup_reduce", [SameOperandsAndResultType]
-    of:
+    this op in convergence.
+
+    If a `cluster_size` is not provided, the reduction covers all lanes of the
+    subgroup and the result is equal for all lanes.
+
+    If a `cluster_size` is provided, the subgroup is divided into clusters of
+    `cluster_size` contiguous lanes each, a reduction is done for all lanes of
+    each cluster (in parallel), and the result is equal for all lanes in a
+    cluster.
```
I'm wondering if I should say instead that the cluster size is the workgroup size if it's not specified.

andfau-amd replied:
Proposed alternative wording:

> The subgroup is divided into clusters of `cluster_size` contiguous lanes each, and a reduction is done for all lanes of each cluster (in parallel). The result is equal for all lanes in a cluster. When `cluster_size` is omitted, there is a single cluster with size of the subgroup.

andfau-amd replied:
I think the description ends up more cohesive this way.

andfau-amd committed:
[mlir][gpu] Add 'cluster_size' attribute to gpu.subgroup_reduce

This enables performing several reductions in parallel, each smaller
than the size of the subgroup.

One potential application is flash attention with subgroup-wide matrix
multiplication and reduction combined in one kernel. The multiplication
operation requires a 2D matrix to be distributed over the lanes of the
subgroup, which then constrains the shape the following reduction can
have if we want to keep data in registers.

Groverkss reviewed (approved):
LGTM thanks!
